File: rabbitmq-jpa-kotlin/domain/src/main/code/sample/domain/model/BaseEntity.java
package sample.domain.model;

import com.google.common.base.MoreObjects;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.DynamicInsert;
import org.hibernate.annotations.DynamicUpdate;
import org.hibernate.envers.Audited;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.Date;

@Getter
@MappedSuperclass
@DynamicInsert
@DynamicUpdate
@EntityListeners(AuditingEntityListener.class)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public abstract class BaseEntity {

    @Column(nullable = false, insertable = true, updatable = false, columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP")
    @CreatedDate
    @CreationTimestamp
    private LocalDateTime createdAt;

    @Column(insertable = true, updatable = true, columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP")
    @CreatedDate
    @LastModifiedDate
    private LocalDateTime updatedAt;

//    @Column(nullable = false, updatable = false)
//    @Temporal(TemporalType.TIMESTAMP)
//    private Date createdAt;
//
//    //audit
//    @Column(nullable = false)
//    @Temporal(TemporalType.TIMESTAMP)
//    private Date lastUpdatedAt;
//
//    @PrePersist
//    protected void onCreate() {
//        createdAt = lastUpdatedAt = new Date();
//    }
//
//    @PreUpdate
//    protected void onUpdate() {
//        lastUpdatedAt = new Date();
//    }

    @Override
    public String toString() {
        return this.buildStringHelper().toString();
    }

    protected MoreObjects.ToStringHelper buildStringHelper() {
        return MoreObjects.toStringHelper(this);
    }
}

File: 2021-rabbitmq-jpa/domain/jpa/src/main/kotlin/sample/domain/jpa/model/BaseEntity.kt
package sample.domain.jpa.model

import com.google.common.base.MoreObjects
import org.hibernate.annotations.ColumnDefault
import org.hibernate.annotations.DynamicInsert
import org.hibernate.annotations.DynamicUpdate
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime
import javax.persistence.Column
import javax.persistence.EntityListeners
import javax.persistence.MappedSuperclass

@MappedSuperclass
@DynamicInsert
@DynamicUpdate
@EntityListeners(AuditingEntityListener::class)
abstract class BaseEntity {
    @Column(
        nullable = false,
        insertable = true,
        updatable = false,
//        columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"
    )
    @CreatedDate //    @CreationTimestamp
    @ColumnDefault("CURRENT_TIMESTAMP")
    val createdAt: LocalDateTime = LocalDateTime.now()

    @Column(
        insertable = true,
        updatable = true,
//        columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP"
    )
    // @CreatedDate
    @LastModifiedDate
    @ColumnDefault("CURRENT_TIMESTAMP")
    val updatedAt: LocalDateTime = LocalDateTime.now()

    //    @Column(nullable = false, updatable = false)
    //    @Temporal(TemporalType.TIMESTAMP)
    //    private Date createdAt;
    //
    //    //audit
    //    @Column(nullable = false)
    //    @Temporal(TemporalType.TIMESTAMP)
    //    private Date lastUpdatedAt;
    //
    //    @PrePersist
    //    protected void onCreate() {
    //        createdAt = lastUpdatedAt = new Date();
    //    }
    //
    //    @PreUpdate
    //    protected void onUpdate() {
    //        lastUpdatedAt = new Date();
    //    }
    override fun toString(): String {
        return buildStringHelper().toString()
    }

    protected fun buildStringHelper(): MoreObjects.ToStringHelper {
        return MoreObjects.toStringHelper(this)
    }
}
